### ISSUE TITLE
Type Inference Fails on Beta and Nightly but not on Stable

### ISSUE THREAD

**Issue by sunjay** [rust-lang/rust#40662]
I setup Travis to run on both beta and nightly. This is recommended by the Travis docs because it helps catch regressions.

For some reason, [my build](https://travis-ci.org/brain-lang/brain/builds/212857183) works on stable, but fails on beta and nightly. Is this a regression?

Here's the code that doesn't work:
```rust
    scope.declare_builtin_function(
        type_name.clone().concat(Identifier::from("writeln")),
        ItemType::Function {
            args: vec![
                FuncArgType::Arg(stdout_type),
                FuncArgType::Variadic(None),
            ],
            return_type: unit_type,
        },
        move |scope, args, _| {
            let mut ops = expression::call(scope, Expression::Identifier(write_method_name.clone()), args)?;

            // Write a newline using a temporary cell
            //TODO: ops.extend(...);
            unimplemented!();

            Ok(ops)
        }
    );
```
This is in [stdio.rs](https://github.com/brain-lang/brain/blob/4a4596b9213e283762473fba467f80c13c3f646a/src/core/io/stdio.rs#L66).

For some reason on beta and nightly the compiler can no longer infer that type:
```
error[E0282]: type annotations needed
  --> src/core/io/stdio.rs:82:13
   |
82 |             Ok(ops)
   |             ^^ cannot infer type for `E`
error: aborting due to previous error
```

I think it should be able to figure it out based on the call to `scope.declare_builtin_function()`.

**Comment by sunjay** [comment]
This seems to have everything to do with the `unimplemented!()` call. Removing that makes it compile successfully on nightly. I still think this is a bug though, since I shouldn't have to remove that. It works in stable after all.

**Comment by sinkuu** [comment]
https://github.com/rust-lang/rust/issues/39984?

**Comment by sunjay** [comment]
@sinkuu This is probably the same as that one. I'll let someone who is sure close the issue.

**Comment by nikomatsakis** [comment]
I can validate that this is fixed by #40636

**Comment by nikomatsakis** [comment]
Closing as a dup of #39984
